File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

// Write your first then apply function here
export function firstThenApply<T, R>(
  a: T[], 
  p: (element: T) => boolean, 
  f: (element: T) => R
): R | undefined {
  const foundElement: T | undefined = a.find(p);
  return foundElement !== undefined ? f(foundElement) : undefined;
}

// Write your powers generator here
export function* powersGenerator(base: bigint): Generator<bigint, void, unknown> {
  let power = 1n;
  
  while (true) {
    yield power;
    power *= base;
  }
}

// Write your line count function here
import { createReadStream } from "fs";
import * as readline from "readline";

export async function meaningfulLineCount(filePath: string): Promise<number> {
  let count = 0;

  // Create read stream and line reader interface
  const fileStream = createReadStream(filePath, { encoding: "utf-8" });

  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  });

  // Read lines asynch
  try {
    for await (const line of rl) {
      const trimmedLine = line.trim();
      if (trimmedLine.length > 0 && !trimmedLine.startsWith("#")) {
        count++;
      }
    }
  } catch (err) {
    throw new Error(`Error reading file: ${err}`);
  } finally {
    rl.close(); // Ensure the readline interface is closed
  }

  return count;
}


// Write your shape type and associated functions here
export type Shape = 
  | { kind: "Sphere"; radius: number }
  | { kind: "Box"; width: number; length: number; depth: number };

// Volume
export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return (4 / 3) * Math.PI * Math.pow(shape.radius, 3);
    case "Box":
      return shape.width * shape.length * shape.depth;
  }
}

// Surface area
export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Sphere":
      return 4 * Math.PI * Math.pow(shape.radius, 2);
    case "Box":
      const { width, length, depth } = shape;
      return 2 * (width * length + length * depth + width * depth);
  }
}


// Write your binary search tree implementation here
export interface BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T>;
  contains(value: T): boolean;
  size(): number;
  inorder(): Generator<T>;
}

// Empty
export class Empty<T> implements BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T> {
    return new Node(value, new Empty(), new Empty());
  }

  contains(_: T): boolean {
    return false;
  }

  size(): number {
    return 0;
  }

  *inorder(): Generator<T> {
  }

  toString(): string {
    return "()";
  }
}

// Node
class Node<T> implements BinarySearchTree<T> {
  constructor(
    private readonly value: T,
    private readonly left: BinarySearchTree<T>,
    private readonly right: BinarySearchTree<T>
  ) {}

  insert(newValue: T): BinarySearchTree<T> {
    if (newValue < this.value) {
      return new Node(this.value, this.left.insert(newValue), this.right);
    } else if (newValue > this.value) {
      return new Node(this.value, this.left, this.right.insert(newValue));
    } else {
      return this;
    }
  }

  contains(target: T): boolean {
    if (target < this.value) {
      return this.left.contains(target);
    } else if (target > this.value) {
      return this.right.contains(target);
    } else {
      return true;
    }
  }

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  *inorder(): Generator<T> {
    yield* this.left.inorder();
    yield this.value;
    yield* this.right.inorder();
  }

  toString(): string {
    const leftStr = this.left instanceof Empty ? "" : `${this.left}`;
    const rightStr = this.right instanceof Empty ? "" : `${this.right}`;
    return `(${leftStr}${this.value}${rightStr})`.trim();
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

(* Write your first then apply function here *)
let rec first_then_apply lst p f =
  match lst with
  | [] -> None
  | x :: xs -> if p x then f x else first_then_apply xs p f

(* Write your powers generator here *)
let powers_generator x =
  let rec generate_power n () =
    Seq.Cons (n, generate_power (n * x))
  in
  generate_power 1

(* Write your line count function here *)
let meaningful_line_count filename =
  let in_channel = open_in filename in
  Fun.protect
    ~finally:(fun () -> close_in in_channel)
    (fun () ->
      let rec count_lines acc =
        match input_line in_channel with
        | line ->
            let trimmed = String.trim line in
            if trimmed = "" || (String.length trimmed > 0 && trimmed.[0] = '#') then
              count_lines acc
            else
              count_lines (acc + 1)
        | exception End_of_file -> acc
      in
      count_lines 0)

(* Write your shape type and associated functions here *)
let pi = 4.0 *. atan 1.0

type shape =
  | Sphere of float
  | Box of float * float * float

let volume s =
  match s with
  | Sphere r -> (4.0 /. 3.0) *. pi *. (r ** 3.0)
  | Box (l, w, h) -> l *. w *. h

let surface_area s =
  match s with
  | Sphere r -> 4.0 *. pi *. (r ** 2.0)
  | Box (l, w, h) -> 2.0 *. (l *. w +. w *. h +. h *. l)
  
(* Write your binary search tree implementation here *)
type 'a bst =
  | Empty
  | Node of 'a bst * 'a * 'a bst

let rec insert (x : 'a) (tree : 'a bst) : 'a bst =
  match tree with
  | Empty -> Node (Empty, x, Empty)
  | Node (left, value, right) ->
      if x < value then
        Node (insert x left, value, right)
      else if x > value then
        Node (left, value, insert x right)
      else
        tree

let rec contains (x : 'a) (tree : 'a bst) : bool =
  match tree with
  | Empty -> false
  | Node (left, value, right) ->
      if x = value then true
      else if x < value then contains x left
      else contains x right

let rec size (tree : 'a bst) : int =
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right

let rec inorder (tree : 'a bst) : 'a list =
  match tree with
  | Empty -> []
  | Node (left, value, right) -> inorder left @ [value] @ inorder right